cmd/githubgen/main.go

```diff
@@ -0,0 +1,208 @@
+
+## 
+# NOTE: New components MUST have a codeowner. Add new components to the CODEOWNERS file instead of here.
+##
+
+## COMMON & SHARED components
+internal/common
+
+`
+
+const unmaintainedStatus = "unmaintained"
+
+// Generates files specific to Github according to status metadata:
+// .github/CODEOWNERS
+// .github/ALLOWLIST
+func main() {
+	flag.Parse()
+	folder := flag.Arg(0)
+	if err := run(folder); err != nil {
+		log.Fatal(err)
+	}
+}
+
+type Codeowners struct {
+	// Active codeowners
+	Active []string `mapstructure:"active"`
+	// Emeritus codeowners
+	Emeritus []string `mapstructure:"emeritus"`
+}
+type Status struct {
+	Stability     map[string][]string `mapstructure:"stability"`
+	Distributions []string            `mapstructure:"distributions"`
+	Class         string              `mapstructure:"class"`
+	Warnings      []string            `mapstructure:"warnings"`
+	Codeowners    *Codeowners         `mapstructure:"codeowners"`
+}
+type metadata struct {
+	// Type of the component.
+	Type string `mapstructure:"type"`
+	// Type of the parent component (applicable to subcomponents).
+	Parent string `mapstructure:"parent"`
+	// Status information for the component.
+	Status *Status `mapstructure:"status"`
+}
+
+func loadMetadata(filePath string) (metadata, error) {
+	cp, err := fileprovider.New().Retrieve(context.Background(), "file:"+filePath, nil)
+	if err != nil {
+		return metadata{}, err
+	}
+
+	conf, err := cp.AsConf()
+	if err != nil {
+		return metadata{}, err
+	}
+
+	md := metadata{}
+	if err := conf.Unmarshal(&md); err != nil {
+		return md, err
+	}
+
+	return md, nil
+}
+
+func run(folder string) error {
+	components := map[string]metadata{}
+	foldersList := []string{}
+	maxLength := 0
+	err := filepath.Walk(folder, func(path string, info fs.FileInfo, err error) error {
+		if info.Name() == "metadata.yaml" {
+			m, err := loadMetadata(path)
+			if err != nil {
+				return err
+			}
+			if m.Status == nil {
+				return nil
+			}
+			key := filepath.Dir(path) + "/"
+			components[key] = m
+			foldersList = append(foldersList, key)
+			for stability := range m.Status.Stability {
+				if stability == unmaintainedStatus {
+					// do not account for unmaintained status to change the max length of the component line.
+					return nil
+				}
+			}
+			if len(key) > maxLength {
+				maxLength = len(key)
+			}
+		}
+		return nil
+	})
+	sort.Strings(foldersList)
+	if err != nil {
+		return err
+	}
+	codeowners := codeownersHeader
+	deprecatedList := "## DEPRECATED components\n"
+	unmaintainedList := "\n## UNMAINTAINED components\n"
+
+	unmaintainedCodeowners := unmaintainedHeader
+	currentFirstSegment := ""
+LOOP:
+	for _, key := range foldersList {
+		m := components[key]
+		for stability := range m.Status.Stability {
+			if stability == unmaintainedStatus {
+				unmaintainedList += key + "\n"
+				unmaintainedCodeowners += fmt.Sprintf("%s%s @open-telemetry/collector-contrib-approvers \n", key, strings.Repeat(" ", maxLength-len(key)))
```

Comment:
This might just be a personal nit, but there are a few really long lines in this file. I'd prefer to have them split to be more manageable, but realize there's no explicit limit.